doc/TclDouble.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH TclDouble 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclDouble -- This class is used to create, read, and modify double
objects from Java code.

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR(double \fIdoubleValue\fR)
.sp
static double \fBget\fR(Interp \fIinterp\fR, TclObject \fItobj\fR) throws TclException
.sp
static void \fBset\fR(TclObject \fItobj\fR, double \fIdoubleValue\fR)

.SH ARGUMENTS
.AS TclObject "doubleValue" .
.AP double doubleValue
New value for internal double representation.
.AP Interp interp
If an error occurs while converting an object to be a double object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
TclObject to manipulate.  If the internal representation of
\fItobj\fR is not TclDouble, the method converts it to TclDouble.
.BE

.SH DESCRIPTION
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new TclObject whose internal
representation is the double \fIdoubleValue\fR.  The precision of
\fIdoubleValue\fR is controlled by the global Tcl variable
tcl_precision, whose default value is 12.

.TP
\fBget\fR
The \fBget\fR method returns a double value from the \fItobj\fR.  If an error
occurs while converting \fItobj\fR to be a double object, an error
message is left in the \fIinterp\fR's result object unless
\fIinterp\fR is NULL.
.TP
\fBset\fR
The \fBset\fR method sets \fItobj\fR to have the double value of \fIdoubleValue\fR.
When this method returns, the internal representation of \fItobj\fR
is TclDouble.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewDoubleObj, Tcl_GetDoubleFromObj, Tcl_SetDoubleObj

.SH "SEE ALSO"
InternalRep, TclObject, TclList, TclIndex, TclString, TclInteger,
TclBoolean, ReflectObject

.SH KEYWORDS
double, internal representation, object, object type, precision, tcl_precision


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

src/tcljava/tcl/lang/TclDouble.java
/*
 * TclDouble.java --
 *
 *	Implements the TclDouble internal object representation, as well
 *	variable traces for the tcl_precision variable.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclDouble.java,v 1.8 2006/05/15 01:25:46 mdejong Exp $
 *
 */

package tcl.lang;

/*
 * This class implements the double object type in Tcl.
 */

public class TclDouble implements InternalRep {

/*
 * Internal representation of a double value.
 */

private double value;

// Extra debug checking

private final static boolean validate = false;


/*
 *----------------------------------------------------------------------
 *
 * TclDouble --
 * 
 *	Construct a TclDouble representation with the given double
 *	value.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private
TclDouble(
    double d)		// Initial value.
{
    value = d;

    if (TclObject.saveObjRecords) {
        String key = "TclDouble";
        Integer num = (Integer) TclObject.objRecordMap.get(key);
        if (num == null) {
            num = new Integer(1);
        } else {
            num = new Integer(num.intValue() + 1);
        }
        TclObject.objRecordMap.put(key, num);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * TclDouble --
 *
 *	Construct a TclDouble representation with the initial value
 *	taken from the given string.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private
TclDouble(
    Interp interp,		// Current interpreter.
    String str)			// String that contains the initial value.
throws
    TclException		// If error occurs in string conversion.
{
    value = Util.getDouble(interp, str);

    if (TclObject.saveObjRecords) {
        String key = "TclDouble";
        Integer num = (Integer) TclObject.objRecordMap.get(key);
        if (num == null) {
            num = new Integer(1);
        } else {
            num = new Integer(num.intValue() + 1);
        }
        TclObject.objRecordMap.put(key, num);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * duplicate --
 *
 *	Duplicate the current object.
 *
 * Results:
 *	A dupilcate of the current object.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

public InternalRep duplicate()
{
    if (TclObject.saveObjRecords) {
        String key = "TclDouble.duplicate()";
        Integer num = (Integer) TclObject.objRecordMap.get(key);
        if (num == null) {
            num = new Integer(1);
        } else {
            num = new Integer(num.intValue() + 1);
        }
        TclObject.objRecordMap.put(key, num);
    }

    return new TclDouble(value);
}

/**
 * Implement this no-op for the InternalRep interface.
 */

public void dispose() {}

/*
 *----------------------------------------------------------------------
 *
 * newInstance --
 *
 *	Creates a new instance of a TclObject with a TclDouble internal
 *	representation.
 *
 * Results:
 *	The newly created TclObject.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

public static TclObject
newInstance(
    double d)			// Initial value.
{
    return new TclObject(new TclDouble(d));
}

/*
 *----------------------------------------------------------------------
 *
 * setDoubleFromAny --
 *
 *	Called to convert a TclObject's internal rep to TclDouble.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	When successful, the internal representation of tobj is
 *	changed to TclDouble, if it is not already so.
 *
 *----------------------------------------------------------------------
 */

private static void
setDoubleFromAny(
    Interp interp,		// Current interpreter. May be null.
    TclObject tobj)		// The object to convert.
throws
    TclException		// If error occurs in type conversion.
				// Error message will be left inside
				// the interp if it's not null.

{
    // This method is only ever invoked from TclDouble.get().
    // This method will never be invoked when the internal
    // rep is already a TclDouble. This method will always
    // reparse a double from the string rep so that tricky
    // special cases like "040" are handled correctly.

    if (validate) {
        if (tobj.getInternalRep() instanceof TclDouble) {
            throw new TclRuntimeError("should not be TclDouble, was a " +
                tobj.getInternalRep().getClass().getName());
        }
    }

    tobj.setInternalRep(new TclDouble(interp, tobj.toString()));

    if (TclObject.saveObjRecords) {
        String key = "TclString -> TclDouble";
        Integer num = (Integer) TclObject.objRecordMap.get(key);
        if (num == null) {
            num = new Integer(1);
        } else {
            num = new Integer(num.intValue() + 1);
        }
        TclObject.objRecordMap.put(key, num);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * get --
 *
 *	Returns the double value of the object.
 *
 * Results:
 *	The double value of the object.
 *
 * Side effects:
 *	When successful, the internal representation of tobj is
 *	changed to TclDouble, if it is not already so.
 *
 *----------------------------------------------------------------------
 */

public static double
get(
    Interp interp,		// Current interpreter. May be null.
    TclObject tobj)		// The object to query.
throws
    TclException		// If the object does not have a TclDouble
				// representation and a conversion fails.
				// Error message will be left inside
				// the interp if it's not null.
{
    TclDouble tdouble;

    if (! tobj.isDoubleType()) {
	if (Util.isJacl()) {
	    // Try to convert to TclDouble. If the string can't be
	    // parsed as a double, then raise a TclException here.

	    setDoubleFromAny(interp, tobj);
	    double dval;
	    tdouble = (TclDouble) tobj.getInternalRep();
	    dval = tdouble.value;

	    // The string can be parsed as a double, but if it
	    // can also be parsed as an integer then we need
	    // to convert the internal rep back to TclInteger.
	    // This logic handles the special case of an octal
	    // string like "040". The most common path through
	    // this code is a normal double like "1.0", so this
	    // code will only attempt a conversion to TclInteger
	    // when the string looks like an integer. This logic
	    // is tricky, but it leads to a speedup in performance
	    // critical expr code since the double value from a
	    // TclDouble can be used without having to check to
	    // see if the double looks like an integer.

	    if (Util.looksLikeInt(tobj.toString())) {
	        try {
	            int ival = TclInteger.get(null, tobj);

	            // A tricky octal like "040" can be parsed as
	            // the double 40.0 or the integer 32, return
	            // the value parsed as a double and leave
	            // the object with a TclInteger internal rep.

	            return dval;
	        } catch (TclException te) {
	            throw new TclRuntimeError("looksLikeInt() is true, " +
                        "but TclInteger.get() failed for \"" +
                        tobj.toString() + "\"");
	        }
	    }

	    if (validate) {
	        // Double check that we did not just create a TclDouble
	        // that looks like an integer.

	        InternalRep tmp = tobj.getInternalRep();
	        if (!(tmp instanceof TclDouble)) {
                    throw new TclRuntimeError("not a TclDouble, is a " +
                        tmp.getClass().getName());
	        }
	        String stmp = tobj.toString();
                if (Util.looksLikeInt(stmp)) {
                    throw new TclRuntimeError("looks like an integer");
	        }
	    }
        } else {
	    setDoubleFromAny(interp, tobj);
	    tdouble = (TclDouble) tobj.getInternalRep();
        }
    } else {
	tdouble = (TclDouble) tobj.getInternalRep();
    }

    return tdouble.value;
}

/*
 *----------------------------------------------------------------------
 *
 * set --
 *
 *	Changes the double value of the object.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The internal representation of tobj is
 *	changed to TclDouble, if it is not already so.
 *
 *----------------------------------------------------------------------
 */

public static void
set(
    TclObject tobj,		// The object to modify.
    double d)			// The new value for the object. 
{
    tobj.invalidateStringRep();

    if (tobj.isDoubleType()) {
	TclDouble tdouble = (TclDouble) tobj.getInternalRep();
	tdouble.value = d;
    } else {
	tobj.setInternalRep(new TclDouble(d));
    }
}

/*
 *----------------------------------------------------------------------
 *
 * toString --
 *
 *	Called to query the string representation of the Tcl
 *	object. This method is called only by TclObject.toString()
 *	when TclObject.stringRep is null.
 *
 * Results:
 * 	Returns the string representation of the TclDouble object.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

public String
toString()
{
    return Util.printDouble(value);
}

/**
 * This special helper method is used only by
 * the Expression module. This method will
 * change the internal rep to a TclDouble with
 * the passed in double value. This method does
 * not invalidate the string rep since the
 * object's value is not being changed.
 *
 * @param tobj the object to operate on.
 * @param d the new double value.
 */
static void exprSetInternalRep(TclObject tobj, double d) {
    if (validate) {

        // Double check that the internal rep is not
        // already of type TclDouble.

        InternalRep rep = tobj.getInternalRep();

        if (rep instanceof TclDouble) {
            throw new TclRuntimeError("exprSetInternalRep() called with object" +
                " that is already of type TclDouble");
        }

        // Double check that the new int value and the
        // string rep would parse to the same integer.

        double d2;
        try {
            d2 = Util.getDouble(null, tobj.toString());
        } catch (TclException te) {
            throw new TclRuntimeError("exprSetInternalRep() called with double" +
                " value that could not be parsed from the string");
        }
        if (d != d2) {
            throw new TclRuntimeError("exprSetInternalRep() called with double value " +
                d + " that does not match parsed double value " + d2 +
                ", parsed from str \"" +
                tobj.toString() + "\"");
        }

        // It should not be possible to parse the TclObject's string
        // rep as an integer since we know it is a double. An object
        // that could be parsed as either a double or an integer
        // should have been parsed as an integer.

        try {
            int ival = Util.getInt(null, tobj.toString());
            throw new TclRuntimeError("should not be able to parse string rep as int: " +
                tobj.toString());
        } catch (TclException e) {
            // No-op
        }
    }

    tobj.setInternalRep(new TclDouble(d));
}

} // end TclDouble

